File: Cocoa/AppFrameworks/UIKit/CollectionViewManager/CollectionCell/YJUICollectionViewCell.m
//
//  YJUICollectionViewCell.m
//  YJUICollectionView
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 16/5/21.
//  Copyright © 2016年 YJCocoa. All rights reserved.
//

#import "YJUICollectionViewCell.h"
#import "YJUICollectionViewManager.h"
#import "YJNSFoundationOther.h"

@implementation UICollectionViewCell (YJUICollectionView)

+ (CGSize)collectionViewManager:(YJUICollectionViewManager *)collectionViewManager sizeForCellObject:(YJUICollectionCellObject *)cellObject {
    if (cellObject.createCell == YJUICollectionCellCreateXib) { // 默认使用xib创建cell
        NSArray<UIView *> *array = [[NSBundle mainBundle] loadNibNamed:cellObject.cellName owner:nil options:nil];
        return array.firstObject.frame.size;
    }
    return collectionViewManager.delegateFlowLayoutManager.flowLayout.itemSize; // 默认设置
}

@end

@implementation YJUICollectionViewCell

#pragma mark getter & setter
- (NSString *)reuseIdentifier {
    NSString *reuseIdentifier = [super reuseIdentifier];
    if (reuseIdentifier) return reuseIdentifier;
    return YJNSStringFromClass(self.class);
}

#pragma mark - 
+ (YJUICollectionCellCreate)cellCreate {
    if ([YJNSStringFromClass([YJUICollectionViewCell class]) isEqualToString:YJNSStringFromClass(self.class)]) {
        return YJUICollectionCellCreateClass;
    }
    return [super cellCreate];
}

- (void)reloadDataSyncWithCellObject:(YJUICollectionCellObject *)cellObject collectionViewManager:(nonnull YJUICollectionViewManager *)collectionViewManager {
    _cellObject = cellObject;
    _collectionViewManager = collectionViewManager;
}

@end

File: YJCocoa/Classes/AppFrameworks/UIKit/CollectionView/YJUICollectionViewCell.swift
//
//  YJUICollectionViewCell.swift
//  YJCocoa
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 2019/5/23.
//  Copyright © 2016-现在 YJCocoa. All rights reserved.
//

import UIKit

@objc
extension UICollectionViewCell {
    
    /// 获取 cell 的显示size
    open class func collectionViewManager(_ collectionViewManager: YJUICollectionViewManager, sizeWith cellObject: YJUICollectionCellObject) -> CGSize {
        return collectionViewManager.flowLayout.itemSize
    }
    
}

@objcMembers
open class YJUICollectionViewCell: UICollectionViewCell {
    
    public private(set) var cellObject: YJUICollectionCellObject!
    public private(set) var collectionViewManager: YJUICollectionViewManager!
    
    override open func collectionViewManager(_ collectionViewManager: YJUICollectionViewManager, reloadWith cellObject: YJUICollectionCellObject) {
        super.collectionViewManager(collectionViewManager, reloadWith: cellObject)
        self.collectionViewManager = collectionViewManager
        self.cellObject = cellObject
    }
    
}
